File: app/src/main/java/org/checklist/comics/comicschecklist/ui/ComicAdapter.java
package org.checklist.comics.comicschecklist.ui;

import android.databinding.DataBindingUtil;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.checklist.comics.comicschecklist.R;
import org.checklist.comics.comicschecklist.databinding.ListItemViewBinding;
import org.checklist.comics.comicschecklist.model.Comic;

import java.util.List;

public class ComicAdapter extends RecyclerView.Adapter<ComicAdapter.ComicViewHolder> {

    List<? extends Comic> mComicList;

    @Nullable
    private final ComicClickCallback mComicClickCallback;

    ComicAdapter(@Nullable ComicClickCallback clickCallback) {
        mComicClickCallback = clickCallback;
    }

    public void setComicList(final List<? extends Comic> comicList) {
        if (mComicList == null) {
            mComicList = comicList;
            notifyItemRangeInserted(0, comicList.size());
        } else {
            DiffUtil.DiffResult result = DiffUtil.calculateDiff(new ComicDiffCallback((List<Comic>) comicList, (List<Comic>) mComicList));
            mComicList = comicList;
            result.dispatchUpdatesTo(this);
        }
    }

    @Override
    public ComicViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ListItemViewBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.list_item_view,
                        parent, false);
        binding.setCallback(mComicClickCallback);
        return new ComicViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(ComicViewHolder holder, int position) {
        holder.binding.setComic(mComicList.get(position));
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mComicList == null ? 0 : mComicList.size();
    }

    static class ComicViewHolder extends RecyclerView.ViewHolder {

        final ListItemViewBinding binding;

        public ComicViewHolder(ListItemViewBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/org/checklist/comics/comicschecklist/ui/ComicAdapter.kt
package org.checklist.comics.comicschecklist.ui

import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import org.checklist.comics.comicschecklist.R
import org.checklist.comics.comicschecklist.database.entity.ComicEntity
import org.checklist.comics.comicschecklist.databinding.ListItemViewBinding

internal class ComicAdapter internal constructor(private val mComicClickCallback: ComicClickCallback?) : androidx.recyclerview.widget.RecyclerView.Adapter<ComicAdapter.ComicViewHolder>() {

    internal lateinit var mComicList: List<ComicEntity>

    fun setComicList(comicList: List<ComicEntity>) {
        if (!::mComicList.isInitialized) {
            mComicList = comicList
            notifyItemRangeInserted(0, comicList.size)
        } else {
            val result = DiffUtil.calculateDiff(ComicDiffCallback(comicList, mComicList))
            mComicList = comicList
            result.dispatchUpdatesTo(this)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ComicViewHolder {
        val binding = DataBindingUtil
                .inflate<ListItemViewBinding>(LayoutInflater.from(parent.context), R.layout.list_item_view,
                        parent, false)
        binding.callback = mComicClickCallback
        return ComicViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ComicViewHolder, position: Int) {
        holder.binding.comic = mComicList[position]
        holder.binding.executePendingBindings()
    }

    override fun getItemCount(): Int {
        return mComicList.size
    }

    internal class ComicViewHolder(val binding: ListItemViewBinding) : androidx.recyclerview.widget.RecyclerView.ViewHolder(binding.root)
}
